dashboard/app/controllers/pd/workshop_daily_survey_controller.rb

```diff
@@ -225,6 +225,38 @@ def new_post
       render :new_general
     end
 
+    # GET workshop_survey/csf/pre201
+    def new_csf_pre201
+      workshop = Pd::Workshop.
+        where(course: COURSE_CSF, subject: SUBJECT_CSF_201).
+        enrolled_in_by(current_user).nearest
+
+      return render :not_enrolled unless workshop
+      return render :too_late unless workshop.state != STATE_ENDED
+      render_csf_survey(PRE_DEEPDIVE_SURVEY, workshop)
+    end
+
+    # GET workshop_survey/csf/post201(/:enrollment_code)
+    def new_csf_post201
```

Comment:
Does this route require the user to be signed-in? I don't see tests covering the signed-out case.  I thought the reason for passing the enrollment code here was to let a teacher fill out the survey without signing in, but you're still calling `current_user` down in the `render_csf_survey` method, assuming it will be set.

Reply:
I think all actions require logging-in in the current setting because of `authorize_resource class: 'Pd::Enrollment'` at the beginning of the controller. This change doesn't support filling out survey without logging in yet. In addition, enrollment code alone, which is a 10-char code, is not designed to and not enough to uniquely identify an user.

Reply:
Got it.  Maybe good enough for now to add tests showing login is required then, and we'll update them if we ever change that behavior.

Reply:
I will add one

Reply:
We can exempt a specific route from the `authorize_resource` call by using `:except` ([documentation](https://www.rubydoc.info/github/CanCanCommunity/cancancan/CanCan%2FControllerAdditions%2FClassMethods:authorize_resource)) ([example](https://github.com/code-dot-org/code-dot-org/blob/staging/dashboard/app/controllers/courses_controller.rb#L4))

Reply:
Thanks Clare, this will be helpful.

Reply:
Also - I think the [enrollment code is unique](https://github.com/code-dot-org/code-dot-org/blob/21b4463a622f42569f1ec5350355aef6f84bc171/dashboard/app/models/pd/enrollment.rb#L351), and we can get the user from the enrollment (if there is a user associated - CSF enrollments may not have a user), so it should be enough to identify the user.